compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -3863,20 +3924,168 @@ private void validateWorkerInteractions(WorkerActionSystem workerActionSystem) {
+                }
+            }
+        }
+        return isValid;
+    }
+
+    private void handleWaitAction(WorkerActionSystem workerActionSystem, BLangNode currentAction,
+                                  WorkerActionStateMachine worker) {
+        if (currentAction instanceof BLangWaitForAllExpr) {
+            boolean allWorkersAreDone = true;
+            BLangWaitForAllExpr waitForAllExpr = (BLangWaitForAllExpr) currentAction;
+            for (BLangWaitForAllExpr.BLangWaitKeyValue keyValuePair : waitForAllExpr.keyValuePairs) {
+                BSymbol workerSymbol = getWorkerSymbol(keyValuePair);
+                if (isWorkerSymbol(workerSymbol)
+                        && isWorkerFromFunction(workerActionSystem.getActionEnvironment(currentAction), workerSymbol)) {
+                    var otherSM = workerActionSystem.find(workerSymbol.name.value);
+                    allWorkersAreDone = allWorkersAreDone && otherSM.done();
+                }
+            }
+            if (allWorkersAreDone) {
+                worker.next();
+                this.workerSystemMovementSequence++;
+            }
+        } else {
+            BLangWaitExpr wait = (BLangWaitExpr) currentAction;
+            List<String> workerNameList = getWorkerNameList(wait.exprList.get(0),
+                    workerActionSystem.getActionEnvironment(currentAction));
+            if (workerNameList.isEmpty()) {
+                // No workers found, there must be only future references in the waiting list, we can move to next state
+                worker.next();
+                this.workerSystemMovementSequence++;
+            }
+            for (String workerName : workerNameList) {
+                // If any worker in wait is done, we can continue.
+                var otherSM = workerActionSystem.find(workerName);
+                if (otherSM.done()) {
+                    worker.next();
+                    this.workerSystemMovementSequence++;
+                    break;
+                }
+            }
+        }
+    }
+
+    private BSymbol getWorkerSymbol(BLangWaitForAllExpr.BLangWaitKeyValue keyValuePair) {
+        BLangExpression value = keyValuePair.getValue();
+        if (value != null && value.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+            return ((BLangSimpleVarRef) value).symbol;
+        } else if (keyValuePair.keyExpr != null && keyValuePair.keyExpr.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+            return ((BLangSimpleVarRef) keyValuePair.keyExpr).symbol;
+        }
+        return null;
+    }
+
+    private List<String> getWorkerNameList(BLangExpression expr, SymbolEnv functionEnv) {
+        ArrayList<String> workerNames = new ArrayList<>();
+        populateWorkerNameList(expr, workerNames, functionEnv);
+        return workerNames;
+    }
+
+    private void populateWorkerNameList(BLangExpression expr, ArrayList<String> workerNames, SymbolEnv functionEnv) {
+        if (expr.getKind() == NodeKind.BINARY_EXPR) {
```

Comment:
what is the case where we get a Binary expression?

Reply:
It's alternate wait action, I think I have missed to add negative tests for them, will add (`wait w|v`)

Reply:
yes, let's add that as a separate PR. I will merge this for now. Please create an issue for that.

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/30285